patroni/postgresql.py

```diff
@@ -655,12 +750,84 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
-    def restart(self):
-        self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting():
+            return not self.check_startup_state_changed()
         else:
+            ready = self.pg_isready()
```

Comment:
Ah ok, that makes sense. One point of view would be that if the system is so slow that `pg_ctl status` makes ttl run out, then losing the master key is the desired outcome as the system is not really fit to be running a database. The watchdog patch I'm polishing up right now will help with the issue of PostgreSQL still running as master after the lock times out. Probably an additional layer of health checks would be nice too to detect almost-but-not-completely-dead infrastructure. It's hard to do anything useful about not having enough CPU slices to run trivial programs or not enough memory to start PG backends. Except raising our hands up, saying that we can't work like this and shutting everything down in the hope that someone else is in a better position.

Do you want to take a shot at moving pg_isready -> state=starting after a failed is_leader check or should I take it on?